File: sandbox/plc4go/internal/plc4go/modbus/readwrite/ModbusTcpADU.go
//
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//
package readwrite

import (
	"errors"
	"plc4x.apache.org/plc4go-modbus-driver/0.8.0/internal/plc4go/spi"
	"reflect"
	"strconv"
)

// Constant values.
const PROTOCOLIDENTIFIER uint16 = 0x0000

// The data-structure of this message
type ModbusTcpADU struct {
	transactionIdentifier uint16
	unitIdentifier        uint8
	pdu                   ModbusPDU
}

// The corresponding interface
type IModbusTcpADU interface {
	spi.Message
	Serialize(io spi.WriteBuffer)
}

func NewModbusTcpADU(transactionIdentifier uint16, unitIdentifier uint8, pdu ModbusPDU) spi.Message {
	return &ModbusTcpADU{transactionIdentifier: transactionIdentifier, unitIdentifier: unitIdentifier, pdu: pdu}
}

func (m ModbusTcpADU) LengthInBits() uint16 {
	var lengthInBits uint16 = 0

	// Simple field (transactionIdentifier)
	lengthInBits += 16

	// Const Field (protocolIdentifier)
	lengthInBits += 16

	// Implicit Field (length)
	lengthInBits += 16

	// Simple field (unitIdentifier)
	lengthInBits += 8

	// Simple field (pdu)
	lengthInBits += m.pdu.LengthInBits()

	return lengthInBits
}

func (m ModbusTcpADU) LengthInBytes() uint16 {
	return m.LengthInBits() / 8
}

func ModbusTcpADUParse(io spi.ReadBuffer, response bool) (spi.Message, error) {

	// Simple Field (transactionIdentifier)
	var transactionIdentifier uint16 = io.ReadUint16(16)

	// Const Field (protocolIdentifier)
	var protocolIdentifier uint16 = io.ReadUint16(16)
	if protocolIdentifier != PROTOCOLIDENTIFIER {
		return nil, errors.New("Expected constant value " + strconv.Itoa(int(PROTOCOLIDENTIFIER)) + " but got " + strconv.Itoa(int(protocolIdentifier)))
	}

	// Implicit Field (length) (Used for parsing, but it's value is not stored as it's implicitly given by the objects content)
	var _ uint16 = io.ReadUint16(16)

	// Simple Field (unitIdentifier)
	var unitIdentifier uint8 = io.ReadUint8(8)

	// Simple Field (pdu)
	_pduMessage, _err := ModbusPDUParse(io, bool(response))
	if _err != nil {
		return nil, errors.New("Error parsing simple field 'pdu'. " + _err.Error())
	}
	var pdu ModbusPDU
	pdu, _pduOk := _pduMessage.(ModbusPDU)
	if !_pduOk {
		return nil, errors.New("Couldn't cast message of type " + reflect.TypeOf(_pduMessage).Name() + " to ModbusPDU")
	}

	// Create the instance
	return NewModbusTcpADU(transactionIdentifier, unitIdentifier, pdu), nil
}

func (m ModbusTcpADU) Serialize(io spi.WriteBuffer) {
	serializeFunc := func(typ interface{}) {
		if _, ok := typ.(IModbusPDU); ok {

			// Simple Field (transactionIdentifier)
			var transactionIdentifier uint16 = m.transactionIdentifier
			io.WriteUint16(16, (transactionIdentifier))

			// Const Field (protocolIdentifier)
			io.WriteUint16(16, 0x0000)

			// Implicit Field (length) (Used for parsing, but it's value is not stored as it's implicitly given by the objects content)
			var length uint16 = ((m.pdu.LengthInBytes()) + (1))
			io.WriteUint16(16, (length))

			// Simple Field (unitIdentifier)
			var unitIdentifier uint8 = m.unitIdentifier
			io.WriteUint8(8, (unitIdentifier))

			// Simple Field (pdu)
			var pdu ModbusPDU = m.pdu
			pdu.Serialize(io)
		}
	}
	serializeFunc(m)
}

File: sandbox/plc4py/plc4py/protocols/umas/readwrite/ModbusTcpADU.py
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#

from dataclasses import dataclass

from plc4py.api.exceptions.exceptions import PlcRuntimeException
from plc4py.api.exceptions.exceptions import SerializationException
from plc4py.api.messages.PlcMessage import PlcMessage
from plc4py.protocols.umas.readwrite.ModbusPDU import ModbusPDU
from plc4py.spi.generation.ReadBuffer import ReadBuffer
from plc4py.spi.generation.WriteBuffer import WriteBuffer
from plc4py.utils.GenericTypes import ByteOrder
import math


@dataclass
class ModbusTcpADU:
    transaction_identifier: int
    unit_identifier: int
    pdu: ModbusPDU
    # Arguments.
    umas_request_function_key: int
    PROTOCOL_IDENTIFIER: int = 0x0000

    def serialize(self, write_buffer: WriteBuffer):
        write_buffer.push_context("ModbusTcpADU")

        # Simple Field (transactionIdentifier)
        write_buffer.write_unsigned_short(
            self.transaction_identifier, logical_name="transactionIdentifier"
        )

        # Const Field (protocolIdentifier)
        write_buffer.write_unsigned_short(
            self.PROTOCOL_IDENTIFIER, logical_name="protocolIdentifier"
        )

        # Implicit Field (length) (Used for parsing, but its value is not stored as it's implicitly given by the objects content)
        length: int = self.pdu.length_in_bytes() + int(1)
        write_buffer.write_unsigned_short(length, logical_name="length")

        # Simple Field (unitIdentifier)
        write_buffer.write_unsigned_byte(
            self.unit_identifier, logical_name="unitIdentifier"
        )

        # Simple Field (pdu)
        write_buffer.write_serializable(self.pdu, logical_name="pdu")

        write_buffer.pop_context("ModbusTcpADU")

    def length_in_bytes(self) -> int:
        return int(math.ceil(float(self.length_in_bits() / 8.0)))

    def length_in_bits(self) -> int:
        length_in_bits: int = 0
        _value: ModbusTcpADU = self

        # Simple field (transactionIdentifier)
        length_in_bits += 16

        # Const Field (protocolIdentifier)
        length_in_bits += 16

        # Implicit Field (length)
        length_in_bits += 16

        # Simple field (unitIdentifier)
        length_in_bits += 8

        # Simple field (pdu)
        length_in_bits += self.pdu.length_in_bits()

        return length_in_bits

    @staticmethod
    def static_parse(read_buffer: ReadBuffer, **kwargs):
        if kwargs is None:
            raise PlcRuntimeException(
                "Wrong number of arguments, expected 1, but got None"
            )

        umas_request_function_key: int = 0
        if isinstance(kwargs.get("umas_request_function_key"), int):
            umas_request_function_key = int(kwargs.get("umas_request_function_key"))
        elif isinstance(kwargs.get("umas_request_function_key"), str):
            umas_request_function_key = int(
                str(kwargs.get("umas_request_function_key"))
            )
        else:
            raise PlcRuntimeException(
                "Argument 0 expected to be of type int or a string which is parseable but was "
                + kwargs.get("umas_request_function_key").getClass().getName()
            )

        return ModbusTcpADU.static_parse_context(read_buffer, umas_request_function_key)

    @staticmethod
    def static_parse_context(read_buffer: ReadBuffer, umas_request_function_key: int):
        read_buffer.push_context("ModbusTcpADU")

        transaction_identifier: int = read_buffer.read_unsigned_short(
            logical_name="transactionIdentifier",
            bit_length=16,
            byte_order=ByteOrder.BIG_ENDIAN,
            umas_request_function_key=umas_request_function_key,
        )

        PROTOCOL_IDENTIFIER: int = read_buffer.read_unsigned_short(
            logical_name="protocolIdentifier",
            byte_order=ByteOrder.BIG_ENDIAN,
            umas_request_function_key=umas_request_function_key,
        )

        length: int = read_buffer.read_unsigned_short(
            logical_name="length",
            byte_order=ByteOrder.BIG_ENDIAN,
            umas_request_function_key=umas_request_function_key,
        )

        unit_identifier: int = read_buffer.read_unsigned_byte(
            logical_name="unitIdentifier",
            bit_length=8,
            byte_order=ByteOrder.BIG_ENDIAN,
            umas_request_function_key=umas_request_function_key,
        )

        pdu: ModbusPDU = read_buffer.read_complex(
            read_function=ModbusPDU.static_parse,
            logical_name="pdu",
            byte_order=ByteOrder.BIG_ENDIAN,
            umas_request_function_key=umas_request_function_key,
        )

        read_buffer.pop_context("ModbusTcpADU")
        # Create the instance
        _modbus_tcp_adu: ModbusTcpADU = ModbusTcpADU(
            transaction_identifier, unit_identifier, pdu, umas_request_function_key
        )
        return _modbus_tcp_adu

    def equals(self, o: object) -> bool:
        if self == o:
            return True

        if not isinstance(o, ModbusTcpADU):
            return False

        that: ModbusTcpADU = ModbusTcpADU(o)
        return (
            (self.transaction_identifier == that.transaction_identifier)
            and (self.unit_identifier == that.unit_identifier)
            and (self.pdu == that.pdu)
            and True
        )

    def hash_code(self) -> int:
        return hash(self)

    def __str__(self) -> str:
        pass
        # write_buffer_box_based: WriteBufferBoxBased = WriteBufferBoxBased(True, True)
        # try:
        #    write_buffer_box_based.writeSerializable(self)
        # except SerializationException as e:
        #    raise PlcRuntimeException(e)

        # return "\n" + str(write_buffer_box_based.get_box()) + "\n"
